File: audiobook/src/main/java/de/ph1b/audiobook/adapter/BookmarkAdapter.java
package de.ph1b.audiobook.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.ph1b.audiobook.R;
import de.ph1b.audiobook.interfaces.ForApplication;
import de.ph1b.audiobook.model.Bookmark;
import de.ph1b.audiobook.model.Chapter;
import de.ph1b.audiobook.utils.App;

/**
 * Adapter for displaying a list of bookmarks.
 *
 * @author Paul Woitaschek
 */
public class BookmarkAdapter extends RecyclerView.Adapter<BookmarkAdapter.ViewHolder> {

    @NonNull
    private final List<Bookmark> bookmarks;
    private final List<Chapter> chapters;
    @NonNull
    private final OnOptionsMenuClickedListener listener;
    @ForApplication
    @Inject
    Context c;


    public BookmarkAdapter(@NonNull List<Bookmark> bookmarks, List<Chapter> chapters, @NonNull OnOptionsMenuClickedListener listener) {
        App.getComponent().inject(this);
        this.bookmarks = new ArrayList<>(bookmarks);
        this.chapters = new ArrayList<>(chapters);
        this.listener = listener;
    }

    @NonNull
    private static String formatTime(int ms) {
        String h = String.valueOf(TimeUnit.MILLISECONDS.toHours(ms));
        String m = String.format("%02d", (TimeUnit.MILLISECONDS.toMinutes(ms) % 60));
        String s = String.format("%02d", (TimeUnit.MILLISECONDS.toSeconds(ms) % 60));
        String returnString = "";
        if (!h.equals("0")) {
            returnString += h + ":";
        }
        returnString += m + ":" + s;
        return returnString;
    }

    public void removeItem(Bookmark bookmark) {
        int index = bookmarks.indexOf(bookmark);
        bookmarks.remove(bookmark);
        notifyItemRemoved(index);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate
                (R.layout.dialog_bookmark_row_layout, parent, false);
        return new ViewHolder(v, listener);
    }

    public void bookmarkUpdated(Bookmark oldBookmark, Bookmark newBookmark) {
        int oldIndex = bookmarks.indexOf(oldBookmark);
        bookmarks.set(oldIndex, newBookmark);
        notifyItemChanged(oldIndex);
        Collections.sort(bookmarks);
        notifyItemMoved(oldIndex, bookmarks.indexOf(newBookmark));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Bookmark bookmark = bookmarks.get(position);
        holder.title.setText(bookmark.title());

        int size = chapters.size();
        Chapter currentChapter = null;
        for (Chapter c : chapters) {
            if (c.file().equals(bookmark.mediaFile())) {
                currentChapter = c;
            }
        }
        if (currentChapter == null) {
            throw new IllegalArgumentException("Current chapter not found with bookmark=" + bookmark);
        }
        int index = chapters.indexOf(currentChapter);

        holder.summary.setText(c.getString(R.string.format_bookmarks_n_of, index + 1, size));
        holder.time.setText(c.getString(R.string.format_bookmarks_time, formatTime(bookmark.time()),
                formatTime(currentChapter.duration())));
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return bookmarks.size();
    }

    public interface OnOptionsMenuClickedListener {
        void onOptionsMenuClicked(Bookmark bookmark, View v);

        void onBookmarkClicked(Bookmark bookmark);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        @Bind(R.id.edit) ImageButton imageButton;
        @Bind(R.id.text1) TextView title;
        @Bind(R.id.text2) TextView summary;
        @Bind(R.id.text3) TextView time;

        public ViewHolder(View itemView, final OnOptionsMenuClickedListener listener) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onBookmarkClicked(bookmarks.get(getAdapterPosition()));
                }
            });
        }

        @OnClick(R.id.edit)
        void optionsMenuClicked() {
            listener.onOptionsMenuClicked(bookmarks.get(getAdapterPosition()), imageButton);
        }
    }
}

File: audiobook/src/main/kotlin/de/ph1b/audiobook/adapter/BookmarkAdapter.kt
package de.ph1b.audiobook.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import de.ph1b.audiobook.R
import de.ph1b.audiobook.injection.App
import de.ph1b.audiobook.model.Bookmark
import de.ph1b.audiobook.model.Chapter
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * Adapter for displaying a list of bookmarks.

 * @author Paul Woitaschek
 */
class BookmarkAdapter(bookmarks: List<Bookmark>, chapters: List<Chapter>, private val listener: BookmarkAdapter.OnOptionsMenuClickedListener) : RecyclerView.Adapter<BookmarkAdapter.ViewHolder>() {

    private val bookmarks: MutableList<Bookmark>
    private val chapters: List<Chapter>

    @Inject
    internal lateinit var c: Context


    init {
        App.getComponent().inject(this)
        this.bookmarks = ArrayList(bookmarks)
        this.chapters = ArrayList(chapters)
    }

    private fun formatTime(ms: Int): String {
        val h = TimeUnit.MILLISECONDS.toHours(ms.toLong()).toString()
        val m = "%02d".format((TimeUnit.MILLISECONDS.toMinutes(ms.toLong()) % 60))
        val s = "%02d".format((TimeUnit.MILLISECONDS.toSeconds(ms.toLong()) % 60))
        var returnString = ""
        if (h != "0") {
            returnString += h + ":"
        }
        returnString += m + ":" + s
        return returnString
    }

    fun removeItem(bookmark: Bookmark) {
        val index = bookmarks.indexOf(bookmark)
        bookmarks.remove(bookmark)
        notifyItemRemoved(index)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.dialog_bookmark_row_layout, parent, false)
        return ViewHolder(v, listener)
    }

    fun bookmarkUpdated(oldBookmark: Bookmark, newBookmark: Bookmark) {
        val oldIndex = bookmarks.indexOf(oldBookmark)
        bookmarks[oldIndex] = newBookmark
        notifyItemChanged(oldIndex)
        Collections.sort(bookmarks)
        notifyItemMoved(oldIndex, bookmarks.indexOf(newBookmark))
    }


    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val bookmark = bookmarks[position]
        holder.title.text = bookmark.title

        val size = chapters.size
        var currentChapter = chapters.single { it.file == bookmark.mediaFile }
        val index = chapters.indexOfRaw(currentChapter)

        holder.summary.text = c.getString(R.string.format_bookmarks_n_of, index + 1, size)
        holder.time.text = c.getString(R.string.format_bookmarks_time, formatTime(bookmark.time),
                formatTime(currentChapter.duration))
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItemCount(): Int {
        return bookmarks.size
    }

    interface OnOptionsMenuClickedListener {
        fun onOptionsMenuClicked(bookmark: Bookmark, v: View)

        fun onBookmarkClicked(bookmark: Bookmark)
    }

    internal inner class ViewHolder(itemView: View, listener: OnOptionsMenuClickedListener) : RecyclerView.ViewHolder(itemView) {

        private val imageButton: ImageButton
        internal val title: TextView
        internal val summary: TextView
        internal val time: TextView

        init {
            imageButton = itemView.findViewById(R.id.edit) as ImageButton
            title = itemView.findViewById(R.id.text1) as TextView
            summary = itemView.findViewById(R.id.text2) as TextView
            time = itemView.findViewById(R.id.text3) as TextView

            imageButton.setOnClickListener { listener.onOptionsMenuClicked(bookmarks[adapterPosition], imageButton) }
            itemView.setOnClickListener { listener.onBookmarkClicked(bookmarks[adapterPosition]) }
        }
    }
}